notes/supernova.md

```diff
@@ -0,0 +1,259 @@
+
+Let $F_0, \ldots, F\_{\ell-1}$ be folding circuits with the same arity $a$. These circuits implement the `circuit_supernova::StepCircuit` trait, where the main differences with the existing `StepCircuit` trait are
+- The circuit $F_j$ provides its `circuit_index` $j$
+- The `synthesize` function upon input $z_i$ returns the next `program_counter` $\mathsf{pc}\_{i+1}$ alongside the output $z\_{i+1}$. It also accepts the (optional) input program counter $\mathsf{pc}_i$, which can be `None` when $\ell=1$. During circuit synthesis, a constraint enforces $\mathsf{pc}_i \equiv j$. In contrast to the paper, the _predicate_ function $\varphi$ is built into the circuit itself. In other words, we have the signature $(\mathsf{pc}\_{i+1}, z\_{i+1}) \gets F\_{j}(\mathsf{pc}\_{i}, z\_{i})$.
+
+The goal is to efficiently prove the following computation:
+```
+pc_i = pc_0
+z_i = z_0
+for i in 0..num_steps
+	(pc_i, z_i) = F_{pc_i}(z_i)
+return z_i
+```
+
+## Prover state 
+
+The prover needs to store data about the previous function iteration. It is defined by the `supernova::RecursiveSNARK` struct. It contains:
+- $i$: the number of iterations performed. Note that the `new` constructor actually performs the first iteration, and the first call to `prove_step` simply sets the counter to 1. 
+- Primary curve:
+	- $(\mathsf{pc}_i, z_0,z_i)$: current program counter and inputs for the primary circuit
+	- $U[\ ],W[\ ]$: List of relaxed instance/witness pairs for all the circuits on the primary curve. These can be `None` when the circuit for that pair has not yet been executed. The last updated entry is the result of having folded a proof for the correctness of $z_i$.
+- Secondary curve
+	- $(z_0', z_i')$: Inputs for the single circuit on the secondary curve. 
+	- $u',w'$: Proof for the correctness of the circuit that produced $z_i'$ 
+	- $U', W'$: Relaxed instance/witness pair into which $(u', w')$ will be folded into in the next iteration.
+
+## Prove Step
+
+At each step, the prover needs to:
+- Create a proof $T'$ for folding $u'$ into $U'$, producing $U'\_{next}$
+- Create a proof $(u,w)$ on the primary curve for the statements:
+	- $(\mathsf{pc}\_{i+1}, z\_{i+1}) \gets F_\mathsf{pc_i}(z_i)$
+	- Verifying the folding of $u'$ into $U'$ with $T'$
+- Create a proof $T$ for folding $u$ into $U[\mathsf{pc}_i]$, producing $U\_{next}$ 
+- Create a proof $(u'\_{next}, w'\_{next})$ for the verification on the secondary curve
+	- Verifying the folding of $u$ into $U[\mathsf{pc}_i]$ with $T$
+- Update the state of the claims
+	- $U' = U'\_{next}$, $W' = W'\_{next}$
+	- $U[\mathsf{pc}_i] = U\_{next}$, $W[\mathsf{pc}_i] = W\_{next}$
+	- $(u',w') = (u'\_{next}, w'\_{next})$ 
+- Save $z\_{i+1},z'\_{i+1}, \mathsf{pc}\_{i+1}$ as inputs for the next iteration. 
+
+
+
+In pseudocode, `prove_step` looks something like:
+
+```
+if i = 0 {
```

Comment:
A [Mermaid](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) diagram to show the flow of the protocol might help, perhaps more than pseudocode?